### ISSUE TITLE
Multi-head attention implementation difference in fairseq

### ISSUE THREAD

**Issue by getao** [facebookresearch/fairseq#4383]
Hello,

I found there are two kinds of implementation in fairseq. One is calling F.multi_head_attention_forward() in
https://github.com/pytorch/fairseq/blob/806855bf660ea748ed7ffb42fe8dcc881ca3aca0/fairseq/modules/multihead_attention.py#L292;
The other is the remaining part of the forward function in MultiheadAttention class.
I found these two implementations will generate different training result. And using F.multi_head_attention_forward() will generate more stable training loss curve. 
Could anyone help me understand why F.multi_head_attention_forward() can perform better? If I have to use the other way for multi-head attention implementation, how should I improve?

Thanks
